datafusion/physical-plan/src/windows/mod.rs

```diff
@@ -174,20 +174,15 @@ fn create_built_in_window_expr(
     name: String,
     ignore_nulls: bool,
 ) -> Result<Arc<dyn BuiltInWindowFunctionExpr>> {
-    // need to get the types into an owned vec for some reason
-    let input_types: Vec<_> = args
-        .iter()
-        .map(|arg| arg.data_type(input_schema))
-        .collect::<Result<_>>()?;
+    // derive the output datatype from incoming schema
+    let out_data_type: &DataType = input_schema.field_with_name(&name)?.data_type();
```

Comment:
I am not sure how this works (or makes sense), can you help me understand? How can you decide the output type of a function without knowing what the function is (i.e. using `fun.return_type`)?  Is it always guaranteed that you will find `name` in `input_schema`?  If yes, no problem. But if no, shouldn't you fall back to using `fun.return_type`?

I understand that some people may want to have `ROW_NUMBER` output an `Int` instead of an `UInt` (which doesn't make sense but I can see how such an ask may arise for backwards compatibility reasons).

Reply:
Actually, I think with these changes parameter `input_schema: &Schema,` name is misleading in the `create_built_in_window_expr` function. Similarly, `logical_input_schema: &DFSchema,` parameter for the `create_window_expr` is misleading. Now these parameters are the schema of the window function (not input schema). Hence, I think it is guaranteed to find name of the window function in the schema (As long as window function name calculations are consistent). However, we should change these parameters name I think. I checked, how `LogicalPlan::Window` calculates its schema.
It call `exprlist_to_fields` util to calculate fields of window expressions. 
- Which in turn uses `.to_field` method on `input_schema`
- Which in turn uses `Expr::get_type` method on `input_schema`
- which calculates output type using `fun.return_type(&data_types)`. 

In this sense, I think this PR behaves same with current version, and removes duplication. However, we should change parameter/variable names to reflect which data it actually uses.

Reply:
Thanks @mustafasrepo for clarification, I would also like to add the part were `fun.report_type()` called for window builtin functions in logical planner https://github.com/apache/arrow-datafusion/blob/b87dd6143c2dc089b07f74780bd525c4369e68a3/datafusion/expr/src/expr_schema.rs#L184

So when physical layer gets called all the metadata(names, datatypes) is already known